### ISSUE TITLE
Cluster Jgroups Config on Kubernetes environment

### ISSUE THREAD

**Issue by ernestoalejandrocu** [glowroot/glowroot#970]
Hi guys, mi case is the next, i have a glowroot-central collector deployment, i need increase the numbers of replicas (pods) to hold the metric load ingestion from production fleet. My config is:

1 - this file [https://raw.githubusercontent.com/glowroot/glowroot/v0.13.6/central/src/main/resources/jgroups-tcp.xml] is mapped (mounted) into container
2 - my glowroot-central.properties relevant config is:

jgroups.configurationFile=/usr/share/glowroot-central/jgroups-tcp.xml
jgroups.localAddress=0.0.0.0
jgroups.localPort=7800
jgroups.initialNodes=cassandra-master-headless:7800
jgroups.password=xxxxxxxx

The crashing pod show me the next logs:

2023-01-10 09:39:01.308 INFO  org.glowroot - Glowroot version: 0.14.0-beta.3, built 2022-04-03 01:56:31 +0000
2023-01-10 09:39:01.310 INFO  org.glowroot - Java version: 17.0.2
2023-01-10 09:39:02.222 INFO  org.infinispan.CONTAINER - ISPN000128: Infinispan version: Infinispan 'Triskaidekaphobia' 13.0.8.Final
2023-01-10 09:39:02.326 INFO  org.infinispan.CONTAINER - ISPN000556: Starting user marshaller 'org.infinispan.commons.marshall.JavaSerializationMarshaller'
2023-01-10 09:39:02.591 INFO  org.infinispan.CLUSTER - ISPN000078: Starting JGroups channel `ISPN`
2023-01-10 09:39:02.611 WARN  org.jgroups.stack.Configurator - there are more recent versions of FD present; please upgrade to a newer version of the protocol
2023-01-10 09:39:02.690 WARN  org.jgroups.stack.Configurator - SASL has been deprecated; please upgrade to a newer version of the protocol
2023-01-10 09:39:02.723 ERROR org.glowroot - org.infinispan.commons.CacheConfigurationException: ISPN000085: Error while trying to create a channel using the specified configuration file: /usr/share/glowroot-central/jgroups-tcp.xml
org.infinispan.manager.EmbeddedCacheManagerStartupException: org.infinispan.commons.CacheConfigurationException: ISPN000085: Error while trying to create a channel using the specified configuration file: /usr/share/glowroot-central/jgroups-tcp.xml
	at org.infinispan.manager.DefaultCacheManager.internalStart(DefaultCacheManager.java:770)
	at org.infinispan.manager.DefaultCacheManager.start(DefaultCacheManager.java:734)
	at org.infinispan.manager.DefaultCacheManager.<init>(DefaultCacheManager.java:302)
	at org.infinispan.manager.DefaultCacheManager.<init>(DefaultCacheManager.java:224)
	at org.glowroot.central.util.ClusterManager$ClusterManagerImpl.lambda$new$0(ClusterManager.java:139)
	at org.glowroot.central.util.ClusterManager$ClusterManagerImpl.doWithSystemProperties(ClusterManager.java:245)
	at org.glowroot.central.util.ClusterManager$ClusterManagerImpl.<init>(ClusterManager.java:138)
	at org.glowroot.central.util.ClusterManager$ClusterManagerImpl.<init>(ClusterManager.java:112)
	at org.glowroot.central.util.ClusterManager.create(ClusterManager.java:83)
	at org.glowroot.central.CentralModule.<init>(CentralModule.java:184)
	at org.glowroot.central.CentralModule.create(CentralModule.java:137)
	at org.glowroot.central.CentralModule.create(CentralModule.java:132)
	at org.glowroot.central.Main.start(Main.java:44)
	at org.glowroot.central.Main.main(Main.java:28)
Caused by: org.infinispan.commons.CacheConfigurationException: ISPN000085: Error while trying to create a channel using the specified configuration file: /usr/share/glowroot-central/jgroups-tcp.xml
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.buildChannel(JGroupsTransport.java:652)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.initChannel(JGroupsTransport.java:486)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.start(JGroupsTransport.java:471)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:42)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:27)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:617)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:608)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.access$700(BasicComponentRegistryImpl.java:30)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startDependencies(BasicComponentRegistryImpl.java:635)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:599)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.access$700(BasicComponentRegistryImpl.java:30)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startDependencies(BasicComponentRegistryImpl.java:635)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:599)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.access$700(BasicComponentRegistryImpl.java:30)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.AbstractComponentRegistry.internalStart(AbstractComponentRegistry.java:354)
	at org.infinispan.factories.AbstractComponentRegistry.start(AbstractComponentRegistry.java:250)
	at org.infinispan.manager.DefaultCacheManager.internalStart(DefaultCacheManager.java:766)
	... 13 common frames omitted
Caused by: java.lang.Exception: Could not assign property client_callback_handler_class in SASL, method is setClientCallbackHandlerClass, converted value is org.glowroot.central.SaslCallbackHandler
	at org.jgroups.stack.Configurator.resolveAndInvokePropertyMethod(Configurator.java:775)
	at org.jgroups.stack.Configurator.initializeAttrs(Configurator.java:220)
	at org.jgroups.stack.Configurator.createProtocolsAndInitializeAttrs(Configurator.java:132)
	at org.jgroups.stack.Configurator.setupProtocolStack(Configurator.java:71)
	at org.jgroups.stack.Configurator.setupProtocolStack(Configurator.java:50)
	at org.jgroups.stack.ProtocolStack.setup(ProtocolStack.java:493)
	at org.jgroups.JChannel.init(JChannel.java:957)
	at org.jgroups.JChannel.<init>(JChannel.java:134)
	at org.jgroups.JChannel.<init>(JChannel.java:116)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.buildChannel(JGroupsTransport.java:650)
	... 35 common frames omitted
Caused by: java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.jgroups.stack.Configurator.resolveAndInvokePropertyMethod(Configurator.java:771)
	... 44 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.glowroot.central.SaslCallbackHandler
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at org.jgroups.protocols.SASL.setClientCallbackHandlerClass(SASL.java:79)
	... 49 common frames omitted
Exception in thread "main" org.infinispan.manager.EmbeddedCacheManagerStartupException: org.infinispan.commons.CacheConfigurationException: ISPN000085: Error while trying to create a channel using the specified configuration file: /usr/share/glowroot-central/jgroups-tcp.xml
	at org.infinispan.manager.DefaultCacheManager.internalStart(DefaultCacheManager.java:770)
	at org.infinispan.manager.DefaultCacheManager.start(DefaultCacheManager.java:734)
	at org.infinispan.manager.DefaultCacheManager.<init>(DefaultCacheManager.java:302)
	at org.infinispan.manager.DefaultCacheManager.<init>(DefaultCacheManager.java:224)
	at org.glowroot.central.util.ClusterManager$ClusterManagerImpl.lambda$new$0(ClusterManager.java:139)
	at org.glowroot.central.util.ClusterManager$ClusterManagerImpl.doWithSystemProperties(ClusterManager.java:245)
	at org.glowroot.central.util.ClusterManager$ClusterManagerImpl.<init>(ClusterManager.java:138)
	at org.glowroot.central.util.ClusterManager$ClusterManagerImpl.<init>(ClusterManager.java:112)
	at org.glowroot.central.util.ClusterManager.create(ClusterManager.java:83)
	at org.glowroot.central.CentralModule.<init>(CentralModule.java:184)
	at org.glowroot.central.CentralModule.create(CentralModule.java:137)
	at org.glowroot.central.CentralModule.create(CentralModule.java:132)
	at org.glowroot.central.Main.start(Main.java:44)
	at org.glowroot.central.Main.main(Main.java:28)
Caused by: org.infinispan.commons.CacheConfigurationException: ISPN000085: Error while trying to create a channel using the specified configuration file: /usr/share/glowroot-central/jgroups-tcp.xml
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.buildChannel(JGroupsTransport.java:652)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.initChannel(JGroupsTransport.java:486)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.start(JGroupsTransport.java:471)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:42)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:27)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:617)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:608)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.access$700(BasicComponentRegistryImpl.java:30)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startDependencies(BasicComponentRegistryImpl.java:635)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:599)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.access$700(BasicComponentRegistryImpl.java:30)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startDependencies(BasicComponentRegistryImpl.java:635)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:599)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.access$700(BasicComponentRegistryImpl.java:30)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.AbstractComponentRegistry.internalStart(AbstractComponentRegistry.java:354)
	at org.infinispan.factories.AbstractComponentRegistry.start(AbstractComponentRegistry.java:250)
	at org.infinispan.manager.DefaultCacheManager.internalStart(DefaultCacheManager.java:766)
	... 13 more
Caused by: java.lang.Exception: Could not assign property client_callback_handler_class in SASL, method is setClientCallbackHandlerClass, converted value is org.glowroot.central.SaslCallbackHandler
	at org.jgroups.stack.Configurator.resolveAndInvokePropertyMethod(Configurator.java:775)
	at org.jgroups.stack.Configurator.initializeAttrs(Configurator.java:220)
	at org.jgroups.stack.Configurator.createProtocolsAndInitializeAttrs(Configurator.java:132)
	at org.jgroups.stack.Configurator.setupProtocolStack(Configurator.java:71)
	at org.jgroups.stack.Configurator.setupProtocolStack(Configurator.java:50)
	at org.jgroups.stack.ProtocolStack.setup(ProtocolStack.java:493)
	at org.jgroups.JChannel.init(JChannel.java:957)
	at org.jgroups.JChannel.<init>(JChannel.java:134)
	at org.jgroups.JChannel.<init>(JChannel.java:116)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.buildChannel(JGroupsTransport.java:650)
	... 35 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.jgroups.stack.Configurator.resolveAndInvokePropertyMethod(Configurator.java:771)
	... 44 more
Caused by: java.lang.ClassNotFoundException: org.glowroot.central.SaslCallbackHandler
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at org.jgroups.protocols.SASL.setClientCallbackHandlerClass(SASL.java:79)
	... 49 more

Thanks for any help, please is important to me.

Thanks guys.

**Comment by ernestoalejandrocu** [comment]
I recent research i found this:

https://knowledge.informatica.com/s/article/524385?language=en_US

In this case, whe define a variable on deployment like this:

        - name: MY_POD_IP
          valueFrom:
            fieldRef:
              fieldPath: status.podIP

... and we set on glowroot jvm flags the next:

            - name: GLOWROOT_OPTS
              value: >-
                -Xms4g -Xmx4g -XX:+UseG1GC -Djgroups.bind_addr​=$MY_POD_IP

Not working with this config. Also configuring the jgroups glowroot properties ...

jgroups.localAddress=${env:MY_POD_IP} * NO WORK
jgroups.localAddress=$MY_POD_IP * NO WORK
jgroups.localAddress=$(MY_POD_IP) * NO WORK

**** How i can set the atomic ip addres of the pod?

I cant get it work jgroups cluster, i try tcp and udp, cand you help me?

Sorry for my bad english!

**Comment by ernestoalejandrocu** [comment]
i try this udp config for jgroups https://raw.githubusercontent.com/belaban/JGroups/master/conf/udp.xml, but i have those errors....

[glowroot-674cc48fc8-gxpn4.log](https://github.com/glowroot/glowroot/files/10383675/glowroot-674cc48fc8-gxpn4.log)

**Comment by ernestoalejandrocu** [comment]
Updates.....

I could put to working using these udp config for jgroups, looking for the right version of infinispan on the init logs:

(2023-01-10 11:48:02.798 INFO  org.infinispan.CONTAINER - ISPN000128: Infinispan version: Infinispan 'Triskaidekaphobia' 13.0.8.Final)

[config]
https://raw.githubusercontent.com/infinispan/infinispan/13.0.x/core/src/main/resources/default-configs/default-jgroups-udp.xml

You need to solve the buffers warning from the kernel using this post...

https://access.redhat.com/solutions/190643

But my mainly problem doesnt disappear. The nodes running glowroot, increase their load until colapse.

my glowroot config propertie lines are:

    # default is cassandra.contactPoints=127.0.0.1

    cassandra.contactPoints=cassandra-master


    # default is to connect without credentials

    cassandra.username=cassandra


    # default is to connect without credentials

    cassandra.password=cassandra


    # default is cassandra.keyspace=glowroot

    cassandra.keyspace=


    # default is cassandra.consistencyLevel=QUORUM

    # read and write consistency levels can be set to different values by
    separating the two values with

    # a slash, e.g. cassandra.consistencyLevel=ONE/ANY means use consistency
    level ONE for reads and

    # consistency level ANY for writes

    cassandra.consistencyLevel=


    # default is cassandra.maxConcurrentQueries=1024

    cassandra.maxConcurrentQueries=1024


    # default is grpc.bindAddress=0.0.0.0

    grpc.bindAddress=


    # default is grpc.httpPort=8181

    # set this to "none" in order to not serve gRPC over HTTP

    grpc.httpPort=


    # default is grpc.httpsPort=none

    # set this to a port number in order to serve gRPC over HTTPS on that port

    # the certificate and private key to be used must be placed in the same
    directory as this properties

    # file, with filenames "grpc-cert.pem" and "grpc-key.pem" (or "cert.pem" and
    "key.pem" if sharing

    # the same cert between grpc and the ui), where grpc-cert.pem is a PEM
    encoded X.509 certificate

    # chain, and grpc-key.pem is a PEM encoded PKCS#8 private key without a
    passphrase (for example, a

    # self signed certificate can be generated at the command line meeting the
    above requirements using

    # OpenSSL 1.0.0 or later:

    # "openssl req -new -x509 -nodes -days 365 -out grpc-cert.pem -keyout
    grpc-key.pem")

    grpc.httpsPort=


    # default is ui.bindAddress=0.0.0.0

    ui.bindAddress=


    # default is ui.port=4000

    ui.port=


    # default is ui.https=false

    # set this to "true" to serve the UI over HTTPS

    # the certificate and private key to be used must be placed in the same
    directory as this properties

    # file, with filenames "ui-cert.pem" and "ui-key.pem" (or "cert.pem" and
    "key.pem" if sharing

    # the same cert between the ui and grpc), where ui-cert.pem is a PEM encoded
    X.509 certificate

    # chain, and ui-key.pem is a PEM encoded PKCS#8 private key without a
    passphrase (for example, a

    # self signed certificate can be generated at the command line meeting the
    above requirements using

    # OpenSSL 1.0.0 or later:

    # "openssl req -new -x509 -nodes -days 365 -out ui-cert.pem -keyout
    ui-key.pem")

    ui.https=


    # default is ui.contextPath=/

    # this only needs to be changed if reverse proxying the UI behind a non-root
    context path

    ui.contextPath=


    cassandra.symmetricEncryptionKey=XXXXXXXXXXXXXXXX

    jgroups.configurationFile=/usr/share/glowroot-central/jgroups-udp.xml

    jgroups.symEncryptAlgorithm=AES/CBC/PKCS5Padding

    jgroups.symEncryptKeystoreLocation=/usr/share/glowroot-central/jgroups.keystore

    jgroups.symEncryptKeystorePassword=XXXXXXXXXXXXXXXX

    jgroups.symEncryptKeyAlias=XXXXXXXXXXXXXXXX

    jgroups.symEncryptKeyPassword=XXXXXXXXXXXXXXXX


-----------------

These are my JVM opts for glowroot runtime on the kubernetes deployment....

            - name: GLOWROOT_OPTS
              value: >-
                -Xms4g -Xmx4g -XX:+UseG1GC -Djava.net.preferIPv4Stack=true


Any suggest for tunning and better performance?

**Comment by ernestoalejandrocu** [comment]
Updates ....

with this DNS_PING jgroups config, works like sharm, even the peers list is fully right...

https://raw.githubusercontent.com/infinispan/infinispan/13.0.x/core/src/main/resources/default-configs/default-jgroups-kubernetes.xml

Also you have to define the next config property on the glowroot central file:

jgroups.dns.query=glowroot-headless * this the headless kubernetes service to retrieve all the scaled replicas.

But the high load on the node running glowroot, keep happening....

Any suggestions?

Thanks

**Comment by Nowheresly** [comment]
Hi @ernestoalejandrocu 
I would recommend to have a look at this wiki page 

https://github.com/glowroot/glowroot/wiki/Central-Collector-Cluster

Have you trie the version 0.14.beta3 ?

https://github.com/glowroot/glowroot/releases/tag/v0.14.0-beta.3

**Comment by ernestoalejandrocu** [comment]
> Hi @ernestoalejandrocu I would recommend to have a look at this wiki page
> 
> https://github.com/glowroot/glowroot/wiki/Central-Collector-Cluster
> 
> Have you trie the version 0.14.beta3 ?
> 
> https://github.com/glowroot/glowroot/releases/tag/v0.14.0-beta.3

Hi @Nowheresly, thanks for digging, yes i chequeck that doc: https://github.com/glowroot/glowroot/wiki/Central-Collector-Cluster, for me, some things, doesnt work, in my old post lines i write how fix my issues about clustertering on kubernetes. Im using the 0.14.0-beta.3 too from docker hub. Now my problem is heavy load on the nodes where run replicas of glowroot, i observe high iowait from "top", is recommended use SSD volumes for cassadra backends data filesytem?

Thanks

**Comment by ernestoalejandrocu** [comment]
Recently i update one agent to Glowroot version: 0.14.0-beta.3 of the fleet, what is this message?

ERROR o.g.agent.impl.BytecodeServiceImpl - Could not find TLS ALPN provider; no working netty-tcnative, Conscrypt, or Jetty NPN/ALPN available

Any help?

**Comment by ernestoalejandrocu** [comment]
> Are you using java 8? 
> You should consider using java 11+ with the 0.14.* agent.

i cant use a bigger java version at this time, on this app, is very old, with an old jboss 5 eap. Researching a little bit i found if i increase de java 8 to the release 252, it could work. I am going to test right now.

Thanks @Nowheresly

**Comment by ernestoalejandrocu** [comment]
> Are you using java 8? You should consider using java 11+ with the 0.14.* agent.

Confirmed, using jdk1.8.0_321 is working again!

Thanks!

**Comment by ernestoalejandrocu** [comment]
Guys new issue related to this topic about clustering ...

2023-01-11 16:54:07.172 ERROR o.i.s.OutboundTransferTask - Failed to send entries to node glowroot-6df8fcbb8c-khfgt-33810: ISPN000217: Received exception from glowroot-6df8fcbb8c-khfgt-33810, see cause for remote stack trace
org.infinispan.remoting.RemoteException: ISPN000217: Received exception from glowroot-6df8fcbb8c-khfgt-33810, see cause for remote stack trace
	at org.infinispan.remoting.transport.ResponseCollectors.wrapRemoteException(ResponseCollectors.java:25)
	at org.infinispan.remoting.transport.ValidSingleResponseCollector.withException(ValidSingleResponseCollector.java:37)
	at org.infinispan.remoting.transport.ValidSingleResponseCollector.addResponse(ValidSingleResponseCollector.java:21)
	at org.infinispan.remoting.transport.impl.SingleTargetRequest.addResponse(SingleTargetRequest.java:74)
	at org.infinispan.remoting.transport.impl.SingleTargetRequest.onResponse(SingleTargetRequest.java:44)
	at org.infinispan.remoting.transport.impl.RequestRepository.addResponse(RequestRepository.java:51)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processResponse(JGroupsTransport.java:1490)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1392)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.access$300(JGroupsTransport.java:144)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1562)
	at org.jgroups.JChannel.up(JChannel.java:780)
	at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:913)
	at org.jgroups.protocols.FRAG3.up(FRAG3.java:165)
	at org.jgroups.protocols.FlowControl.up(FlowControl.java:347)
	at org.jgroups.protocols.FlowControl.up(FlowControl.java:347)
	at org.jgroups.protocols.pbcast.GMS.up(GMS.java:876)
	at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:254)
	at org.jgroups.protocols.UNICAST3.deliverMessage(UNICAST3.java:1048)
	at org.jgroups.protocols.UNICAST3.addMessage(UNICAST3.java:771)
	at org.jgroups.protocols.UNICAST3.handleDataReceived(UNICAST3.java:752)
	at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:405)
	at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:592)
	at org.jgroups.protocols.VERIFY_SUSPECT.up(VERIFY_SUSPECT.java:132)
	at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:186)
	at org.jgroups.protocols.FD_SOCK.up(FD_SOCK.java:254)
	at org.jgroups.protocols.MERGE3.up(MERGE3.java:281)
	at org.jgroups.protocols.Discovery.up(Discovery.java:300)
	at org.jgroups.protocols.TP.passMessageUp(TP.java:1400)
	at org.jgroups.util.SubmitToThreadPool$SingleMessageHandler.run(SubmitToThreadPool.java:98)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.infinispan.commons.CacheException: ISPN000936: Class 'org.glowroot.ui.ImmutableSession$SerialForm' blocked by deserialization allow list. Adjust the configuration serialization allow list regular expression to include this class.
	at org.infinispan.marshall.exts.ThrowableExternalizer.readObject(ThrowableExternalizer.java:181)
	at org.infinispan.marshall.exts.ThrowableExternalizer.readObject(ThrowableExternalizer.java:42)
	at org.infinispan.marshall.core.GlobalMarshaller.readWithExternalizer(GlobalMarshaller.java:727)
	at org.infinispan.marshall.core.GlobalMarshaller.readNonNullableObject(GlobalMarshaller.java:708)
	at org.infinispan.marshall.core.GlobalMarshaller.readNullableObject(GlobalMarshaller.java:357)
	at org.infinispan.marshall.core.BytesObjectInput.readObject(BytesObjectInput.java:32)
	at org.infinispan.remoting.responses.ExceptionResponse$Externalizer.readObject(ExceptionResponse.java:49)
	at org.infinispan.remoting.responses.ExceptionResponse$Externalizer.readObject(ExceptionResponse.java:41)
	at org.infinispan.marshall.core.GlobalMarshaller.readWithExternalizer(GlobalMarshaller.java:727)
	at org.infinispan.marshall.core.GlobalMarshaller.readNonNullableObject(GlobalMarshaller.java:708)
	at org.infinispan.marshall.core.GlobalMarshaller.readNullableObject(GlobalMarshaller.java:357)
	at org.infinispan.marshall.core.GlobalMarshaller.objectFromObjectInput(GlobalMarshaller.java:191)
	at org.infinispan.marshall.core.GlobalMarshaller.objectFromByteBuffer(GlobalMarshaller.java:220)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processResponse(JGroupsTransport.java:1482)
	... 25 common frames omitted


Any ideas how i can fixit?

Thanks

**Comment by Nowheresly** [comment]
Hi,

this error is described here:

> Caused by: org.infinispan.commons.CacheException: ISPN000936: Class 'org.glowroot.ui.ImmutableSession$SerialForm' blocked by deserialization allow list. Adjust the configuration serialization allow list regular expression to include this class.

This error has been fixed (issue: #925, PR #926) in the `main` branch and will be included int a future release.
In the mean time, you can build your own glowroot central and be sure to include commit https://github.com/glowroot/glowroot/commit/95b53ed1b398e6859d44530f87b6c6991c484955

**Comment by ernestoalejandrocu** [comment]
Fixed using -Dinfinispan.deserialization.allowlist.regexps=.* as workaround!
